### PR TITLE
feature/signInPageObject with Template

### PR DESCRIPTION
**Pull request by agrant23**
…s to elements 10.15.20

### CODE REVIEW

File: pages/coinbase/signin.py [hunk 1]
```diff
@@ -9,12 +9,12 @@
 class SigninPage:
 
     def __init__(self, driver, drivethrough=True, bypass_auth=True):
-
+        
         # sign in page title: Coinbase - Buy/Sell Digital Currency
-        if not drivethrough: driver.get(config.coinbase_domain + '/signin')
+        if drivethrough: driver.get(config.coinbase_domain + '/signin')     #did you throw in that not to trip me up?
```
Comment (MNadolsky) [inline pages/coinbase/signin.py]: No this is not correct. revert.

I wrote documentation for this which should not be confusing. It's not merged in yet though. check the cleanup branch (it's in code review by the way).

File: pages/coinbase/signin.py [hunk 2]
```diff
@@ -53,7 +53,7 @@ def build_elements(self):
         # NAVIGATION BAR
 
         self.home_link =      Link(self.driver, self.home_link_loc)
-        self.products_menu =  ReactiveMenu(self.driver, self.products_menu_loc)
+        self.products_menu =  ReactiveMenu(self.driver, self.products_menu_loc) #needs a driver.maximize_window() before instantiation to successfully implement this object and the remaining objects in the navigation bar
```
Comment (MNadolsky) [inline pages/coinbase/signin.py]: I know. It doesn't need a maximize window though because it's a resolution issue - if your monitor is small enough maxing the window won't help. We just need to make sure the window is a certain size. But yeah, almost the same thing.

Anyway, please write a bug ticket for this problem.

File: pages/coinbase/signin.py [hunk 3]
```diff
@@ -63,12 +63,12 @@ def build_elements(self):
 
         self.email_field =          Field(self.driver, self.email_field_loc)
         self.pass_field =           Field(self.driver, self.pass_field_loc)
-        self.stay_signed_in_checkbox = ''
+        self.stay_signed_in_checkbox = CheckBox(self.driver, self.stay_signed_in_checkbox_loc)
```
Comment (MNadolsky) [inline pages/coinbase/signin.py]: Don't make new templates. put tickets in for them.

Since we have no functionality for checkbox clicking, just leave it empty and make the ticket.

File: pages/templates/elements.py [hunk 4]
```diff
@@ -64,6 +64,24 @@ def click(self):
             EC.visibility_of_element_located(self.wait_locator))
 
 
+class CheckBox(Element):
```
Comment (MNadolsky) [inline pages/templates/elements.py]: remove. If you want to save, make a branch.

File: pages/coinbase/signin.py [hunk 5]
```diff
@@ -9,12 +9,12 @@
 class SigninPage:
```
Comment (MNadolsky) [inline pages/coinbase/signin.py]: remember the 79 character/line limit

File: pages/coinbase/signin.py [hunk 6]
```diff
@@ -85,22 +85,22 @@ def login(self, bypass_auth=True):
 
         inputs
         -----
-        bypass_auth: bool
+        bypass_auth: bool       #Should this docString be near the __init__ method? since bypas_auth is first given there
```
Comment (MNadolsky) [inline pages/coinbase/signin.py]: This has changed in the cleanup branch in code review.

File: pages/coinbase/signin.py [hunk 7]
```diff
@@ -85,22 +85,22 @@ def login(self, bypass_auth=True):
...
-               paste the cookie list into the variable
+            4. Make a file called cookies in /secure, declare a single variable
+               called coinbase, and copy and paste the cookie list into the variable
             5. Delete any newlines in the middle of the dictionary keys/values
 
             If bypass_auth is False, the login process will require two-factor
             auth.
         """
-
+        #I don't understand why loading the cookies is in this method, when the SigninPage object is instantiated this will already be done, is this to give the option to instantiate signin without loading cookies but to login with it.
```
Comment (MNadolsky) [inline pages/coinbase/signin.py]: I have a ticket for this.

**Review by MNadolsky — CHANGES_REQUESTED** [review]

File: pages/templates/elements.py [hunk 8]
```diff
@@ -175,10 +193,10 @@ def select(self, item_selector):
             item_loc = (by.XPATH, f"//*[.='{item_selector}']")
             item = self.driver.find_element(*item_loc)
 
-        if item.tag_name is not 'a': 
+        if item.tag_name != 'a': 
 
             parent = item.find_element_by_xpath('..')            
-            if parent is 'a': item = parent
+            if parent == 'a': item = parent
```
Comment (agrant23) [inline pages/templates/elements.py]: I did on my first commit of sign in page. This does not affect performance but my VsCode says its  an error. I should have branched this off, my bad. I don't know if you want this change implemented to be considerate to other IDEs

Comment (MNadolsky) [reply pages/templates/elements.py]: It says it's an error but it still runs? What error does it give you?

File: pages/coinbase/signin.py [hunk 9]
```diff
@@ -49,20 +49,19 @@ def __init__(self, driver, spawn=False, bypass_auth=True):
...
-    sign_in_button_loc =            (by.ID, 'signin_button')
-    forgot_password_link_loc =      (by.LINK_TEXT, 'Forgot password?')
-    no_account_link_loc =           (by.LINK_TEXT, "Don't have an account?")
-    privacy_policy_link_loc =       (by.PARTIAL_LINK_TEXT, 'Privacy Policy')
+    email_field_loc =                  (by.ID, 'email')
+    pass_field_loc =                   (by.ID, 'password')
+    keep_me_signed_in_checkbox_loc =   (by.ID, 'stay_signed_in')
+    sign_in_button_loc =               (by.ID, 'signin_button')
+    forgot_password_link_loc =         (by.LINK_TEXT, 'Forgot password?')
+    not_have_account_link_loc =        (by.LINK_TEXT, "Don't have an account?")
```
Comment (MNadolsky) [inline pages/coinbase/signin.py]: no_account_link was better. If you can use exact text (like above, "keep_me_signed_in" isn't FULL text, but it still makes sense so okay), do that, but if exact text doesn't make sense using a description is best.

File: pages/coinbase/signin.py [hunk 10]
```diff
@@ -49,20 +49,19 @@ def __init__(self, driver, spawn=False, bypass_auth=True):
...
+    privacy_policy_link_loc =          (by.PARTIAL_LINK_TEXT, 'Privacy Policy')
     two_factor_link_loc = (
         by.LINK_TEXT, 'Have an issue with 2-factor authentication?')
 
     # MISC
 
-    #sign_in_error_alert_loc = (by.LINK_TEXT, 'Invalid email or password.')
-    sign_in_error_alert_loc = (by.CLASS_NAME, 'alert')
+    sign_in_error_alert_loc = (by.LINK_TEXT, 'Invalid email or password.')
```
Comment (MNadolsky) [inline pages/coinbase/signin.py]: why did you change this?

**Review by MNadolsky — CHANGES_REQUESTED** [review]
resolve issues and please answer my questions so we can figure out what to do.

File: pages/coinbase/signin.py (hunk 9 above)
Comment (agrant23) [reply pages/coinbase/signin.py]: I'm confused which one is better to use?

File: pages/coinbase/signin.py (hunk 10 above)
Comment (agrant23) [reply pages/coinbase/signin.py]: the commented out line didn't belong there, but then I decided it was a better locator than the class name locator.  But upon thinking about it further class name is better than link text so I'll change it back

File: pages/templates/elements.py (hunk 8 above)
Comment (agrant23) [reply pages/templates/elements.py]: SyntaxWarning: "is not" with a literal. Did you mean "!="?
if item.tag_name is not 'a':
SyntaxWarning: "is" with a literal. Did you mean "=="?
 if parent is 'a': item = parent

I shouldn't have called it an error I should have called it a warning.

Comment (MNadolsky) [reply pages/templates/elements.py]: The change is fine. keep it.

File: pages/coinbase/signin.py (hunk 9 above)
Comment (MNadolsky) [reply pages/coinbase/signin.py]: no_account_link_loc was better.

File: pages/coinbase/signin.py [hunk 11]
```diff
@@ -40,61 +41,63 @@ def __init__(self, driver, spawn=False, bypass_auth=True):
 
     # NAVIGATION BAR
 
-    home_link_loc =     (by.CLASS_NAME, 'Header__Logo')
```
Comment (MNadolsky) [inline pages/coinbase/signin.py:43]: your indentation is inconsistent.

Tip: don't change other people's indentation that has already been approved for merge.... you're only changing something that has already been accepted and opening it up for rejection. Just add your own things, only edit if you have to.

Comment (agrant23) [reply pages/coinbase/signin.py:43]: I am trying to match all indentations. Which is why I adjusted those
